### ISSUE TITLE
Regression from 1.1.2 to 1.1.3: "keyword not found in header"

### ISSUE THREAD

**Issue by dstndstn** [esheldon/fitsio#325]
Hi,

I'm trying to read a Pan-STARRS DR2 stacked image and while 1.1.2 works, 1.1.3 and 1.1.4 fail with:

```
  File "/Users/dstn/legacypipe/py/legacypipe/image.py", line 1105, in read_image_header
    self._image_header = self.read_image_fits()[self.hdu].read_header()
  File "/usr/local/lib/python3.8/site-packages/fitsio/hdu/base.py", line 343, in read_header
    return FITSHDR(self.read_header_list())
  File "/usr/local/lib/python3.8/site-packages/fitsio/hdu/base.py", line 358, in read_header_list
    return self._FITS.read_header(self._ext+1)
OSError: FITSIO status = 202: keyword not found in header
```

A small subimage reproduces the error:
http://broiler.astrometry.net/~dstn/temp/sub.fits
ie,

```
> python3 -c "import fitsio; fitsio.read_header('/tmp/sub.fits')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/fitsio/fitslib.py", line 238, in read_header
    return FITSHDR(_fits.read_header(hdunum))
OSError: FITSIO status = 202: keyword not found in header
```

These images don't have an EXTNAME card, but that doesn't seem to be the issue.

fitsverify doesn't complain about the image.

It does have HIERARCH cards.

**Comment by esheldon** [comment]
The pdf https://fits.gsfc.nasa.gov/registry/hierarch/hierarch.pdf makes it clear that there should be no "." in these names.

So this is a malformed header. Did you identify what it was between 1.1.2 and 1.1.3 that causes it to fail?

**Comment by esheldon** [comment]
Ahh, it is because we introduced the replacement of junk characters in header names :)

**Comment by esheldon** [comment]
That was a serious issue because not doing so caused seg faults.

Maybe its a book keeping issue: the character gets replaced but then this is not understood somewhere down stream.

**Comment by dstndstn** [comment]
Yes, that is exactly what happens, the keyword gets altered during header parsing, and then later during header parsing it looks up the value for that keyword using the original name, which then isn't found.

As for whether these headers are legal: in the HIERARCH documentation I guess you're referring to this paragraph,
```
Under the ESO implementation of this convention, each token string that precedes the equals sign must only contain characters that are legal in formal FITS keywords, i.e., the uppercase letters A through Z, the digits 0 through 9, and the hyphen and underscore characters. The tokens may, however, be longer than the 8 character limit of formal FITS keywords.
```
which hints that dots aren't allowed, but not in very strong language.  That is definitely a theme in the FITS extension descriptions -- in my view, not nearly enough MUST and SHALL statements!

Regardless, since "once FITS, always FITS", files like this must still remain parseable -- I think your option are to ignore the HIERARCH cards (because if you didn't subscribe to the HIERARCH convention, these would be legal), or parse 'em one way or another.

Do you have a test case for the situation that was causing segfaults that required the `convert_keyword_to_allowed_ascii` call?  Because my #326 PR passes all the tests, but I certainly wouldn't want to cause a regression on that segfault issue!

Perhaps I could add a slightly less strict version of `convert_keyword_to_allowed_ascii` that also allows dots, for use on HIERARCH cards?

Thanks for your time on this, I truly appreciate all the work you put in on fitsio.  I think it is far and away the best FITS library for python.

**Comment by esheldon** [comment]
Is there a way we can modify the code to do the lookup based on the "fixed" keyword name?

**Comment by esheldon** [comment]
And thank for your kind words, I'm glad its useful to others!

**Comment by beckermr** [comment]
That paragraph says "must only" and so I think is extremely clear that periods are not allowed. 

We could parse periods only as "___p___" and then round trip then properly or something.

**Comment by beckermr** [comment]
Well it's supposed to be a p wrapped in triple underscores but typing is hard today.

**Comment by dstndstn** [comment]
On the other hand, the paragraph starts "Under the ESO implementation of this convention" :)  which makes it not a very strong statement, IMO.

The typical FITS stance would be to be permissive in parsing.  I am happy to put in some work to try to make this work.  My trivial approach in #326 works for my use case, but Erin said that conversion logic was there to avoid a segfault, so it would be very helpful to have an instance that triggers that segfault case.  My feeling would be that in an ideal world, you would allow anything in a HIERARCH that doesn't cause problems elsewhere.

Erin also makes a good point that another approach would be to do the book-keeping so that the later header value lookup is done with the modified keyword key.  I can try to implement that if you don't think being more permissive in the HIERARCH header keywords is acceptable.  I would claim, however, that the current behavior of replacing any dots with "JUNK_______" (by index) is quite surprising to users (it certainly was to me!  Why did FPA.ZP turn into FPAKZP!?!).  Always using underscores would I think be more understandable.

**Comment by beckermr** [comment]
> Always using underscores would I think be more understandable.

Could create key collisions against other keywords. You'd need something unique like `___fitsio___` or w/e.

**Comment by esheldon** [comment]
I didn't write `convert_keyword_to_allowed_ascii` so I have no preferences regarding that code.

**Comment by esheldon** [comment]
Or at least I don't remember writing it :)

**Comment by esheldon** [comment]
OK I did write the ascii cleanup for headers:  https://github.com/esheldon/fitsio/pull/302

(I'm happy to replace `JUNK____...` with anything people want.)

The motivation I think was to prevent a seg fault, and it must have worked for that case.   But it is clear from the code logic that this will fail when the keyword has a numerical value this will fail because it is followed up by a call to fits_read_key and that uses the original keyword name, and would either seg fault anyway if we use the original keyword name, or will not find it if we use the converted name.

We could do a check for when it was converted and in that case return it as a string rather than try to convert to a number

**Comment by esheldon** [comment]
By the way, I was not able to make a test case for the seg fault because python refused to accept the characters embedded in the code; we could try to do this somehow externally...

**Comment by dstndstn** [comment]
Since it's a header-parsing issue, can you add an example failing (tiny)
image to the repo?

**Comment by beckermr** [comment]
A small bit of test data would be fine IMHO. We could also base64 encode the data or something like that.

**Comment by esheldon** [comment]
I'm happy to implement specific suggestions

**Comment by beckermr** [comment]
I can try some stuff if someone sends me an example file with the bad data or points me to one. I don't have a specific suggestion but am more than willing to spend some time on it. :)

**Comment by esheldon** [comment]
@dstndstn had an example here, maybe we could just pull out a header manually https://github.com/esheldon/fitsio/issues/295

**Comment by dstndstn** [comment]
I'm sorry, I didn't realize that *I* had reported (or seconded, at least) that segfault case :)  Here is a tiny version of that file,
http://broiler.astrometry.net/~dstn/temp/bad-tiny.fits (11 kB).
I have not yet tried to rewind fitsio to confirm that before the fix for #295 went in that it still segfaults.

**Comment by beckermr** [comment]
Thank you! I will look at this.

**Comment by dstndstn** [comment]
PS, I confirm that this file segfaults with 1.1.2:
```
> python -c "import fitsio; print(fitsio.__version__); fitsio.read_header('/tmp/bad-tiny.fits')"
1.1.2
Segmentation fault: 11
```

**Comment by beckermr** [comment]
We have an 8.4k image in the repo now, so maybe this 11k one is not so bad, assuming we can redistribute it etc.

**Comment by beckermr** [comment]
This file segfaults on master too BTW.

**Comment by beckermr** [comment]
I mean the `bad-tiny.fits` is still segfaulting despite #302.

**Comment by beckermr** [comment]
@dstndstn can you try the code on this branch?

https://github.com/esheldon/fitsio/pull/330

**Comment by esheldon** [comment]
these fixes are released in 1.1.5
